src/Stratis.Bitcoin.Tests/Consensus/ChainedHeaderTreeTest.cs

```diff
@@ -150,5 +155,96 @@ public void ConnectHeaders_NewAndExistingHeaders_ShouldCreateNewHeaders()
         }
+
+        // Supply headers that we already have, make sure no new ChainedHeaders were created.
+        [Fact]
+        public void ConnectHeaders_SupplyExistingHeaders_DontCreateNewChainHeaders()
+        {
+            var testContext = new TestContext();
+            ChainedHeaderTree cht = testContext.CreateChainedHeaderTree();
+            ChainedHeader chainTip = testContext.ExtendAChain(5);
+            Assert.Empty(cht.GetChainedHeadersByHash());
+
+            cht.Initialize(chainTip, true);
+
+            Assert.Equal(6, cht.GetChainedHeadersByHash().Keys.Count);
+
+            var beforeKeys = new List<uint256>(cht.GetChainedHeadersByHash().Keys);
+
+            List<BlockHeader> listOfExistingHeaders = testContext.ChainedHeaderToList(chainTip, 5);
+            cht.ConnectNewHeaders(1, listOfExistingHeaders);
+            var afterKeys = new List<uint256>(cht.GetChainedHeadersByHash().Keys);
+
+            // check chainedHeadersByHash map before and after adding duplicate headers
+            Assert.Equal(beforeKeys.Count, afterKeys.Count);
+            Assert.True(beforeKeys.All(afterKeys.Contains));
+        }
+
+        // Supply some headers and after that supply some more headers
+        // and make sure that PeerTipsByHash is updated (the total amount of items is the same).
+        [Fact]
+        public void ConnectHeaders_SupplyHeadersThenSupplyMore_PeerTipsByHashShouldBeUpdated()
+        {
+            TestContext testContext = new TestContext();
+            ChainedHeaderTree cht = testContext.CreateChainedHeaderTree();
+
+            var chainTip = testContext.ExtendAChain(10);
+            cht.Initialize(chainTip, true);
+            chainTip.BlockDataAvailability = BlockDataAvailabilityState.BlockAvailable;
+            ChainedHeader newChainTip = testContext.ExtendAChain(10, chainTip);
+
+            var listOfExistingHeaders = testContext.ChainedHeaderToList(chainTip, 10);
+            var listOfNewHeaders = testContext.ChainedHeaderToList(newChainTip, 10);
+
+            testContext.ChainStateMock.Setup(s => s.ConsensusTip).Returns(chainTip);
+            chainTip.BlockValidationState = ValidationState.FullyValidated;
+
+            cht.ConnectNewHeaders(1, listOfExistingHeaders);
+
+            var peerIdsByTipHashBefore = new Dictionary<uint256, HashSet<int>>(cht.GetPeerIdsByTipHash());
+            var chainedHeaderTreeHashesBefore = cht.GetChainedHeadersByHash();
+
+            cht.ConnectNewHeaders(1, listOfNewHeaders);
+
+            var peerIdsByTipHashAfter = cht.GetPeerIdsByTipHash();
+            var chainedHeaderTreeHashesAfter = cht.GetChainedHeadersByHash();
+
+            // Same number of entries as we are reassigning hash
+            Assert.True(chainedHeaderTreeHashesBefore.Count == chainedHeaderTreeHashesAfter.Count );
+
+            // Peer's tip hash has changed
+            Assert.True(peerIdsByTipHashBefore.FirstOrDefault(x => x.Value.Contains(1)).Key !=
```

Comment:
The original description of the test it's little bit outdated, we now have 2 structures that we want to check here - peerIdsByTipHash and peerTipsByPeerId. There should be change in both. In the 1st structure you already check it – the value of the peer identifier should move from one key to another key. In the 2nd structure the value should be changed for the peer.